### PR TITLE
[WIP]Gitlab support in pipeline

### PR DESCRIPTION
**Pull request by gitlawr**
Related issue: https://github.com/rancher/rancher/issues/12608

Types PR: https://github.com/rancher/types/pull/402

### CODE REVIEW

**Comment by Double1996** [comment]
cool

**Comment by Double1996** [comment]
cool

File: pkg/api/customization/pipeline/cluster_pipeline.go [hunk 1]
```diff
@@ -186,6 +168,50 @@ func (h *ClusterPipelineHandler) authapp(apiContext *types.APIContext) error {
...
+	}
+	return nil
+}
+
+func (h *ClusterPipelineHandler) configClusterPipelineClientSecret(clusterPipeline *v3.ClusterPipeline, sourceCodeType string, clientSecret string) {
+	if sourceCodeType == model.GithubType && clusterPipeline.Spec.GithubConfig != nil {
+		clusterPipeline.Spec.GithubConfig.ClientSecret = clientSecret
+	} else if sourceCodeType == model.GitlabType && clusterPipeline.Spec.GitlabConfig != nil {
+		clusterPipeline.Spec.GitlabConfig.ClientSecret = clientSecret
+	}
```
Comment (zionwu) [inline pkg/api/customization/pipeline/cluster_pipeline.go]: should we return error if neither of the cases matches?

File: pkg/api/customization/pipeline/cluster_pipeline.go [hunk 2]
```diff
@@ -203,19 +229,15 @@ func (h *ClusterPipelineHandler) authuser(apiContext *types.APIContext) error {
 		return err
 	}
 
-	if authUserInput.SourceCodeType == "github" && clusterPipeline.Spec.GithubConfig == nil {
```
Comment (zionwu) [inline pkg/api/customization/pipeline/cluster_pipeline.go]: This case is removed then maybe we should return error in configClusterPipelineClientSecret?

File: pkg/controllers/user/pipeline/remote/gitlab/webhook.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	respData, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode > 399 {
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/remote/gitlab/webhook.go]: 399 is a magic number. Maybe we can use `resp.StatusCode >= http.StatusBadRequest`?

File: pkg/controllers/user/pipeline/remote/gitlab/webhook.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	respData, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	if resp.StatusCode > 399 {
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/remote/gitlab/webhook.go]: 399 is a magic number. Maybe we can use resp.StatusCode >= http.StatusBadRequest?

File: pkg/controllers/user/pipeline/remote/gitlab/gitlab.go [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("user-agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36)")
+	resp, err := client.Do(req)
+	if err != nil {
+		logrus.Errorf("Received error from gitlab: %v", err)
+		return resp, err
+	}
+	// Check the status code
+	switch resp.StatusCode {
+	case 200:
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/remote/gitlab/gitlab.go]: http.StatusOK

File: pkg/controllers/user/pipeline/remote/gitlab/gitlab.go [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+	case 201:
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/remote/gitlab/gitlab.go]: http.StatusCreated

File: pkg/api/customization/pipeline/cluster_pipeline.go (hunk 2 above)
Comment (gitlawr) [reply pkg/api/customization/pipeline/cluster_pipeline.go]: I'll add a check here

**Comment by zionwu** [comment]
LGTM

File: pkg/controllers/user/pipeline/hooks/drivers/gitlab.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	if err := utils.ValidPipelineSpec(pipeline.Spec); err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	ref, branch, commit := "", "", ""
+	envVars := map[string]string{}
+	if event == gitlabPushEvent {
+		payload := &gitlab.PushEvent{}
```
Comment (alena1108) [inline pkg/controllers/user/pipeline/hooks/drivers/gitlab.go]: can we perhaps make the logic prettier? By:
* creating a wrapper object with ref, branch, commit, envVars as fields
* Create separate methods for `create<Object>GitlabPushEvent`, `create<Object> gitlabMREvent `, `create<Object> gitlabTagEvent `

what do you think?

File: pkg/controllers/user/pipeline/remote/gitlab/gitlab.go [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+
+func New(pipeline v3.ClusterPipeline) (model.Remote, error) {
+	if pipeline.Spec.GitlabConfig == nil {
+		return nil, errors.New("gitlab is not configured")
+	}
+	remote := &client{
+		ClientID:     pipeline.Spec.GitlabConfig.ClientID,
+		ClientSecret: pipeline.Spec.GitlabConfig.ClientSecret,
+	}
+	if pipeline.Spec.GitlabConfig.Host != "" && pipeline.Spec.GitlabConfig.Host != "gitlab.com" {
```
Comment (alena1108) [inline pkg/controllers/user/pipeline/remote/gitlab/gitlab.go]: gitlab.com is hardcoded in many places in the code, can we put it to a const

**Comment by alena1108** [comment]
Closing as the changes will come as a part of https://github.com/rancher/rancher/pull/14283

**Comment by eagle1981** [comment]
In Rancher v2.0.8 still no changes on .../pipeline page
nogitlab-v208

**Pull request closed without merging**